lib/elixir/lib/string.ex

```diff
@@ -1871,6 +1871,22 @@ defmodule String do
     end
   end
 
+  defguardp replace_invalid_ii_of_iii(i, ii)
+            when (896 <= Bitwise.bor(Bitwise.bsl(i, 6), ii) and
+                    Bitwise.bor(Bitwise.bsl(i, 6), ii) <= 1023) or
+                   (32 <= Bitwise.bor(Bitwise.bsl(i, 6), ii) and
+                      Bitwise.bor(Bitwise.bsl(i, 6), ii) <= 863)
```

Comment:
I believe this mostly (see below) equivalent to:

```elixir
when Bitwise.bor(Bitwise.bsl(i, 6), ii) in 896..1023 or
       Bitwise.bor(Bitwise.bsl(i, 6), ii) in 32..896
```

and in fact it seems it's the same as:

```elixir
when Bitwise.bor(Bitwise.bsl(i, 6), ii) in 32..1023
```

?

That being said ` in ` would emit an additional `is_integer` check and I'm not sure whether it can slow things down.

Reply:
Think you misread the `863` for `896` - case in point why `in` is way more legible 😆

Here's old:
```elixir
  defguardp replace_invalid_ii_of_iii(i, ii) when
    (896 <= Bitwise.bor(Bitwise.bsl(i, 6), ii) and Bitwise.bor(Bitwise.bsl(i, 6), ii) <= 1023)
    or
    (32 <= Bitwise.bor(Bitwise.bsl(i, 6), ii) and Bitwise.bor(Bitwise.bsl(i, 6), ii) <= 863)

  defguardp replace_invalid_ii_of_iv(i, ii) when
    (16 <= Bitwise.bor(Bitwise.bsl(i, 6), ii)) and (Bitwise.bor(Bitwise.bsl(i, 6), ii) <= 271)

  defguardp replace_invalid_iii_of_iv(i, ii, iii) when
    (1024 <= Bitwise.bor(Bitwise.bor(Bitwise.bsl(i, 12), Bitwise.bsl(ii, 6)), iii))
    and
    (Bitwise.bor(Bitwise.bor(Bitwise.bsl(i, 12), Bitwise.bsl(ii, 6)), iii) <= 17407)
```

vs new:
```elixir
defguardp replace_invalid_ii_of_iii(i, ii) when
  Bitwise.bor(Bitwise.bsl(i, 6), ii) in 32..863
  or 
  Bitwise.bor(Bitwise.bsl(i, 6), ii) in 896..1023

defguardp replace_invalid_ii_of_iv(i, ii) when
  Bitwise.bor(Bitwise.bsl(i, 6), ii) in 16..271

defguardp replace_invalid_iii_of_iv(i, ii, iii) when
  Bitwise.bor(Bitwise.bor(Bitwise.bsl(i, 12), Bitwise.bsl(ii, 6)), iii) in 1024..17407
```

I've benchmarked the difference [here](https://github.com/elixir-unicode/unicode/blob/validation-perf-comp/benchee/validation.exs) with these [results](https://gist.github.com/Moosieus/66a9398b693f1296ddbae51cfe236a64). Difference seems negligible.

Reply:
ah yeah, of course. sorry about that. :)